pkg/services/ngalert/schedule/fetcher.go

```diff
@@ -3,15 +3,115 @@ package schedule
 import (
 	"context"
 	"fmt"
+	"sync"
 	"time"
 
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/services/ngalert/metrics"
 	"github.com/grafana/grafana/pkg/services/ngalert/models"
+	"github.com/grafana/grafana/pkg/setting"
 )
 
+// RulesStore is a store that provides alert rules for scheduling
+type RulesStore interface {
+	GetAlertRulesKeysForScheduling(ctx context.Context) ([]models.AlertRuleKeyWithVersion, error)
+	GetAlertRulesForScheduling(ctx context.Context, query *models.GetAlertRulesForSchedulingQuery) error
+}
+
+type FetcherCfg struct {
+	ReloadInterval       time.Duration
+	DisableGrafanaFolder bool
+}
+
+// BackgroundFetcher periodically refreshes rules and folders in the background.
+type BackgroundFetcher struct {
+	cfg       FetcherCfg
+	ticker    *time.Ticker
+	ruleStore RulesStore
+	metrics   *metrics.Scheduler
+	logger    log.Logger
+
+	mu                 sync.Mutex
+	latestRules        []*models.AlertRule
+	latestFolderTitles map[models.FolderKey]string
+}
+
+func NewBackgroundFetcher(cfg FetcherCfg, ruleStore RulesStore, metrics *metrics.Scheduler, logger log.Logger) *BackgroundFetcher {
+	if cfg.ReloadInterval <= 0 {
+		logger.Warn("A nonpositive reload interval was provided, falling back to the default")
+		cfg.ReloadInterval = setting.SchedulerBaseInterval
+	}
+	return &BackgroundFetcher{
+		cfg:                cfg,
+		ticker:             time.NewTicker(cfg.ReloadInterval),
+		ruleStore:          ruleStore,
+		metrics:            metrics,
+		logger:             logger,
+		latestRules:        make([]*models.AlertRule, 0),
+		latestFolderTitles: make(map[models.FolderKey]string),
+	}
+}
+
+// Refresh forces a synchronous refresh and blocks until the refresh is done.
+// This is thread-safe with the periodic background refresh.
+func (f *BackgroundFetcher) Refresh(ctx context.Context) {
+	f.updateSchedulableAlertRules(context.Background())
+}
+
+func (f *BackgroundFetcher) Run(ctx context.Context) error {
+	f.logger.Info("Starting rules fetcher")
+	for {
+		select {
+		case <-f.ticker.C:
+			f.logger.Debug("Refreshing rules from storage")
+			f.updateSchedulableAlertRules(context.Background())
```

Comment:
```suggestion
			f.updateSchedulableAlertRules(ctx)
```

Reply:
This one I debated on...should it be a separate context from the grafana startup context, since it's a background process? :thinking: 

(unrelated, we should probably create a trace at this point...)